src/EntityComponentManager.cc

```diff
@@ -490,11 +527,44 @@ void EntityComponentManager::RequestRemoveEntity(Entity _entity,
 /////////////////////////////////////////////////
 void EntityComponentManager::RequestRemoveEntities()
 {
+  if (this->dataPtr->pinnedEntities.empty())
   {
-    std::lock_guard<std::mutex> lock(this->dataPtr->entityRemoveMutex);
-    this->dataPtr->removeAllEntities = true;
+    {
+      std::lock_guard<std::mutex> lock(this->dataPtr->entityRemoveMutex);
+      this->dataPtr->removeAllEntities = true;
+    }
+    this->RebuildViews();
+  }
+  else
+  {
+    std::unordered_set<Entity> tmpToRemoveEntities;
+
+    // Store the to-be-removed entities in a temporary set so we can
+    // mark each of them to be removed from views that contain them.
+    for (const auto &vertex : this->dataPtr->entities.Vertices())
+    {
+      if (std::find(this->dataPtr->pinnedEntities.begin(),
+                    this->dataPtr->pinnedEntities.end(), vertex.first) ==
+          this->dataPtr->pinnedEntities.end())
+      {
+        tmpToRemoveEntities.insert(vertex.first);
+      }
+    }
+
+    {
+      std::lock_guard<std::mutex> lock(this->dataPtr->entityRemoveMutex);
+      this->dataPtr->toRemoveEntities.insert(tmpToRemoveEntities.begin(),
+          tmpToRemoveEntities.end());
+    }
+
+    for (const auto &removedEntity : tmpToRemoveEntities)
+    {
+      for (auto &view : this->dataPtr->views)
+      {
+        view.second.first->MarkEntityToRemove(removedEntity);
```

Comment:
@adlarkin Can you check if I did this right? The code in Dome/Edifice called `UpdateViews` for each removed entity, but that function has been removed in Fortress, but looking at other places where `UpdateViews` was used, I saw that it was replaced with `MarkEntityToRemove`.

Reply:
This looks correct to me! The following tests would fail if the removal didn't work:

https://github.com/ignitionrobotics/ign-gazebo/blob/3478e80eef05d7e23cd63a8ea3eb9eebe9d3bda0/src/EntityComponentManager_TEST.cc#L776

https://github.com/ignitionrobotics/ign-gazebo/blob/3478e80eef05d7e23cd63a8ea3eb9eebe9d3bda0/src/EntityComponentManager_TEST.cc#L1252

Looking at CI, those tests are passing. So, we should be okay :slightly_smiling_face:

Reply:
Thanks!